### ISSUE TITLE
Freebsd13 "Failed to launch: fork/exec /bin/csh: Setctty set but Ctty not valid in child."

### ISSUE THREAD

**Issue by rphanikrishna** [gravitational/teleport#7009]
### Unable to Connect Freebsd13 

getting an error while connecting the Freebsd13 shell from the teleport cluster, below is the error msg 
"Failed to launch: fork/exec /bin/csh: Setctty set but Ctty not valid in child."

![Screenshot 2021-05-22 at 7 41 58 PM](https://user-images.githubusercontent.com/65988249/119229448-c7856100-bb35-11eb-9b1d-f32d26a8bfca.jpg)

### Reproduction Steps
As minimally and precisely as possible, describe step-by-step how to reproduce the problem.
1.Installed teleport in Freebsd13 with "pkg install teleport"  (teleport version installed is Teleport v4.3.9 git:v4.3.9-0-gfabee242d go1.16.3)
2.was able to register Freebsd13 node through tunnel with the below yaml config 
"Teleport:
 nodename: freeBSD13
 auth_token: <xxxxxxxx>
 auth_servers:
 - abc.xyz.com:3080
 data_dir: /var/lib/teleport
 ca_pin: "XXXXXXXXXXXXXXXXXXX"
proxy_service:
 enabled: "no"
auth_service:
 enabled: "no"
ssh_service:
 enabled: "yes"
https_cert_file: /usr/local/share/certs/ca-root-nss.crt
https_key_file: /usr/local/share/certs/ca-root-nss.crt"

3.Teleport Version in Ubuntu 20.04 cluster:  Teleport v6.1.5 git:v6.1.5-0-gdad2a7b5b go1.15.5
4. Freebsd13 is Homelab Hardware/ self hosted, Ubuntu20.04 server is in cloud

**Comment by rphanikrishna** [comment]
https://go-review.googlesource.com/c/go/+/231638/

**Comment by webvictim** [comment]
Try building Teleport 6.1.5 from source on FreeBSD and see whether it helps. There is a pretty big difference between your auth/proxy server and node versions.

**Comment by rphanikrishna** [comment]
with git:v4.3.9-0-gfabee242d go1.16.3
while executing "make full" in freebsd 
below are the errors 

make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 32: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 35: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 37: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 39: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 41: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 43: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 45: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 46: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 55: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 59: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 63: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 66: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 69: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 72: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 73: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 80: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 81: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 84: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 85: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 91: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 93: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 99: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 137: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 139: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 146: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 149: warning: duplicate script for target "ifneq" ignored
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 138: warning: using previous script for "ifneq" defined here
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 149: warning: duplicate script for target "("",)" ignored
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 138: warning: using previous script for "("",)" defined here
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 149: warning: duplicate script for target "("windows")" ignored
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 138: warning: using previous script for "("windows")" defined here
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 150: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 170: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 173: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 174: warning: duplicate script for target "("",)" ignored
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 138: warning: using previous script for "("",)" defined here
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 174: warning: duplicate script for target "("windows")" ignored
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 138: warning: using previous script for "("windows")" defined here
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 175: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 177: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 413: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 416: warning: duplicate script for target "ifneq" ignored
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 138: warning: using previous script for "ifneq" defined here
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 416: warning: duplicate script for target "("",)" ignored
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 138: warning: using previous script for "("",)" defined here
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 416: warning: duplicate script for target "("windows")" ignored
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 138: warning: using previous script for "("windows")" defined here
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 417: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 561: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 562: warning: duplicate script for target "ifneq" ignored
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 138: warning: using previous script for "ifneq" defined here
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 562: warning: duplicate script for target "("",)" ignored
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 138: warning: using previous script for "("",)" defined here
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 563: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 564: Need an operator
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 565: warning: duplicate script for target "ifneq" ignored
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 138: warning: using previous script for "ifneq" defined here
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 565: warning: duplicate script for target "("",)" ignored
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 138: warning: using previous script for "("",)" defined here
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 565: warning: duplicate script for target "("")" ignored
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 562: warning: using previous script for "("")" defined here
make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 566: Need an operator
make: Fatal errors encountered -- cannot continue

**Comment by webvictim** [comment]
I am afraid running Teleport on FreeBSD is unsupported.

**Comment by rphanikrishna** [comment]
but there is already a FreeBSD port of Teleport v4.3.9

**Comment by webvictim** [comment]
It’s unofficial, we didn’t write it.

**Comment by rphanikrishna** [comment]
@webvictim is teleport team planning to support FreeBSD officially in future ?

**Comment by webvictim** [comment]
No plans currently.

**Comment by fernape** [comment]
> with git:v4.3.9-0-gfabee242d go1.16.3
> while executing "make full" in freebsd
> below are the errors
> 
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 32: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 35: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 37: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 39: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 41: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 43: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 45: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 46: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 55: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 59: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 63: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 66: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 69: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 72: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 73: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 80: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 81: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 84: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 85: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 91: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 93: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 99: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 137: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 139: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 146: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 149: warning: duplicate script for target "ifneq" ignored
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 138: warning: using previous script for "ifneq" defined here
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 149: warning: duplicate script for target "("",)" ignored
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 138: warning: using previous script for "("",)" defined here
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 149: warning: duplicate script for target "("windows")" ignored
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 138: warning: using previous script for "("windows")" defined here
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 150: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 170: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 173: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 174: warning: duplicate script for target "("",)" ignored
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 138: warning: using previous script for "("",)" defined here
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 174: warning: duplicate script for target "("windows")" ignored
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 138: warning: using previous script for "("windows")" defined here
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 175: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 177: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 413: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 416: warning: duplicate script for target "ifneq" ignored
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 138: warning: using previous script for "ifneq" defined here
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 416: warning: duplicate script for target "("",)" ignored
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 138: warning: using previous script for "("",)" defined here
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 416: warning: duplicate script for target "("windows")" ignored
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 138: warning: using previous script for "("windows")" defined here
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 417: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 561: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 562: warning: duplicate script for target "ifneq" ignored
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 138: warning: using previous script for "ifneq" defined here
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 562: warning: duplicate script for target "("",)" ignored
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 138: warning: using previous script for "("",)" defined here
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 563: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 564: Need an operator
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 565: warning: duplicate script for target "ifneq" ignored
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 138: warning: using previous script for "ifneq" defined here
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 565: warning: duplicate script for target "("",)" ignored
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 138: warning: using previous script for "("",)" defined here
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 565: warning: duplicate script for target "("")" ignored
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 562: warning: using previous script for "("")" defined here
> make: "/usr/local/go/bin/src/github.com/gravitational/teleport/Makefile" line 566: Need an operator
> make: Fatal errors encountered -- cannot continue

Please, use gmake to build

**Comment by rphanikrishna** [comment]
@fernape thank you i have tried with gmake and below if the output 

"---> Building OSS binaries.
gmake build/teleport build/tctl build/tsh
gmake[2]: Entering directory '/usr/local/go/bin/src/github.com/gravitational/teleport'
GOOS=freebsd GOARCH=amd64 CGO_ENABLED=1 go build -tags "pam   webassets_embed" -o build/teleport  -ldflags '-w -s' ./tool/teleport
# github.com/flynn/u2f/u2fhid
vendor/github.com/flynn/u2f/u2fhid/hid.go:52:18: undefined: hid.Devices
vendor/github.com/flynn/u2f/u2fhid/hid.go:69:21: info.Open undefined (type *hid.DeviceInfo has no field or method Open)
gmake[2]: *** [Makefile:124: build/teleport] Error 2
gmake[2]: Leaving directory '/usr/local/go/bin/src/github.com/gravitational/teleport'
gmake[1]: *** [Makefile:111: all] Error 2
gmake[1]: Leaving directory '/usr/local/go/bin/src/github.com/gravitational/teleport'
gmake: *** [Makefile:138: full] Error 2"



111:        $(MAKE) $(BINARIES)
124:       GOOS=$(OS) GOARCH=$(ARCH) $(CGOFLAG) go build -tags "$(PAM_TAG) $(FIPS_TAG) $(BPF_TAG) $(WEBASSETS_TAG)" -o $(BUILDDIR)/teleport $(BUILDFLAGS) ./tool/teleport
138:    $(MAKE) all WEBASSETS_TAG="webassets_embed"

**Comment by webvictim** [comment]
It looks like the HID library that Teleport is using for interaction with hardware tokens isn't supported on FreeBSD: https://github.com/flynn/u2f

**Comment by tacerus** [comment]
Hi,

I encounter the same issue. I am aware the pot is not official, however seeing that the Teleport software itself runs, it would be amazing if this library issue could be circumvented.

@webvictim Does this mean it might only be caused by an optional library handling FIDO tokens. which are not necessarily being used?

**Comment by webvictim** [comment]
Yes - Teleport uses the aforementioned U2F library for interacting with hardware devices when signing into Teleport. My understanding is that it's only required for the `tsh` binary (which actually handles `tsh login` and 2-factor tokens) rather than the `teleport` binary, though.

We'd definitely be happy to look at a PR which enables to Teleport to build on FreeBSD.

cc @awly

**Comment by tacerus** [comment]
I tried modifying the Makefile in order for it to only build the `teleport` binary, however it too has references to the U2F library, causing the same error to occur upon trying to build it. If I find a way to remove it I will try to work it into an if-clause - that is if no one else has a quicker way to "skip" the U2F part - I'm not too good with  Go.

**Comment by awly** [comment]
The library is also used in server-side U2F validation: https://github.com/gravitational/teleport/blob/master/lib/auth/u2f/authenticate.go#L27
Unfortunately, it has to be compiled along with `teleport`.

There are 2 options to fix this:
- add freebsd support upstream in https://github.com/flynn/hid
- compile out all the [U2F code](https://github.com/gravitational/teleport/tree/master/lib/auth/u2f) on freebsd using Go build tags and make all U2F operations fail there

I don't think anyone from the core Teleport team will have the time to work on this in the near future.
You're welcome to send a PR though!

**Comment by tacerus** [comment]
I made it work.

Unfortunately I am not good with Go at all, so my solution (after quite some time of trying to study build constraints and build tags) is very unorthodox. After modifying the Makefile to only build the `teleport` binary I replaced `hid.go` with https://github.com/karalabe/hid/blob/master/hid.go, pasted the Device function from the original hid.go and butchered the rest of the HID library as well as the `teleport` source by mercilessly commenting out failing lines and blocks.

Of course this unfortunately neither qualifies for a PR nor for detailed instructions and I would not recommend anyone to do what I did in a production environment, so it's not much help for other Teleport users on BSD.

If I find myself with a more elegant solution one day I will try to contribute it, but for now, it doesn't seem to get any better. Apologies to others finding themselves in the situation.
